src/gleam/uri.gleam

```diff
@@ -164,3 +168,133 @@ pub fn to_string(uri: Uri) -> String {
   |> dynamic.string
   |> result.unwrap("")
 }
+
+/// Fetch the origin of a uri
+///
+/// Return the origin of a uri as defined in
+/// https://tools.ietf.org/html/rfc6454
+///
+/// The supported uri schemes are `http` and `https`
+/// Urls without a scheme will return Error
+pub fn origin(uri: Uri) -> Result(String, Nil) {
+  let Uri(scheme: scheme, host: host, port: port, ..) = uri
+  case scheme {
+    Some("https") | Some("http") -> {
+      let origin = Uri(scheme, None, host, port, "", None, None)
+      Ok(to_string(origin))
+    }
+    _ -> Error(Nil)
+  }
+}
+
+fn option_or(first: Option(a), second: Option(a)) -> Option(a) {
+  case first {
+    Some(_) -> first
+    None -> second
+  }
+}
+
+fn drop_last(elements) {
+  let tuple(keep, _last) = list.split(elements, list.length(elements) - 1)
+  keep
+}
+
+fn join_segments(segments) {
+  case segments {
+    [] -> ""
+    _ -> string.append("/", string.join(segments, "/"))
+  }
+}
+
+/// Resolve a uri with respect to the given base uri
+///
+/// The base uri must be an absolute uri or this function will return an error.
+/// The algorithm for merging uris is described in [RFC 3986](https://tools.ietf.org/html/rfc3986#section-5.2)
+pub fn merge(base: Uri, relative: Uri) -> Result(Uri, Nil) {
```

Comment:
This function is very long, might be possible to shorten if there was a dot syntax for updating records.